### PR TITLE
Eureka integration tests

### PR DESCRIPTION
**Pull request by ijsnellf**
**What this PR does / why we need it**:
Eureka integration tests.

**Which issue this PR fixes**: contributes to https://github.com/istio/pilot/issues/1232

**Special notes for your reviewer**:
I'm looking for quick feedback on the overall approach. When `bin/e2e.sh -registry Eureka` is called, I spin up an Eureka server and an accompanying mirroring process that listens for k8s endpoints and registers them with the Eureka server. Then e2e tests proceed mostly as normal. This approach has the advantage requiring very minimal changes to the e2e tests.

TODO:
- [x] improve synchronization logic in the Eureka mirror.
- [x] fix ingress/egress support (perhaps a follow-on PR)

TODO future PRs:
- remove Eureka client duplication (switch to `hudl/fargo` or another 3rd party Eureka library?)
- enable automated the Eureka tests

Note that normal e2e tests will be unaffected by this PR.

**Release note**:

```release-note
```

### CODE REVIEW

File: docker/Dockerfile.eurekamirror [hunk 1]
```diff
@@ -0,0 +1,3 @@
+FROM scratch
```
Comment (rshriram) [inline docker/Dockerfile.eurekamirror]: Add a comment stating that this container is used only for internal testing and what it does.

File: test/integration/egress_rules.go [hunk 2]
```diff
@@ -39,6 +40,14 @@ func (t *egressRules) setup() error {
 
 // TODO: test negatives
 func (t *egressRules) run() error {
+	if !t.Egress {
+		glog.Info("skipping test since egress is missing")
+		return nil
+	}
+	if platform.ServiceRegistry(t.Registry) == platform.EurekaRegistry { // TODO: re-enable once Eureka egress is working
```
Comment (rshriram) [inline test/integration/egress_rules.go]: This should work out of the box in Eureka as well right? These are egress rules that come directly from CRDs. No service registry lookup.

File: test/integration/ingress.go [hunk 3]
```diff
@@ -39,6 +40,9 @@ func (t *ingress) setup() error {
 	if !t.Ingress {
 		return nil
 	}
+	if platform.ServiceRegistry(t.Registry) == platform.EurekaRegistry { // TODO: re-enable once Eureka ingress is working
+		return nil
+	}
```
Comment (rshriram) [inline test/integration/ingress.go]: Same applies here. Unless the ingress has to know anything about Eureka, this should not be any different. It is referring to services by name. (Imagine there is some foo system where you can create ingress elements, outside k8s. the config would look similar to k8s ingress config). 
Or am I missing something?

File: test/integration/tcp.go [hunk 4]
```diff
@@ -36,6 +37,10 @@ func (t *tcp) teardown() {
 }
 
 func (t *tcp) run() error {
+	// TCP in Eureka is tested by the headless service test.
```
Comment (rshriram) [inline test/integration/tcp.go]: 👍

File: test/integration/testdata/eureka.yaml.tmpl [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: eureka
+    spec:
+      containers:
+      - name: eureka
+        image: netflixoss/eureka:1.3.1
```
Comment (rshriram) [inline test/integration/testdata/eureka.yaml.tmpl]: 👍 nice!

File: test/integration/ingress.go (hunk 3 above)
Comment (ijsnellf) [reply test/integration/ingress.go]: Yes, you are correct. It should be working. I'm still trying to figure out what is going wrong.

**Comment by ijsnellf** [comment]
@rshriram: ingress/egress are currently broken. I'm trying to figure out the cause.

**Comment by kyessenov** [comment]
Nice work! Can you add some comments for the purpose of eurekamirror?

File: test/eurekamirror/eurekamirror.go [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"time"
+
+	log "fmt"
```
Comment (GregHanson) [inline test/eurekamirror/eurekamirror.go]: remove duplicate import and format statments

File: test/integration/egress_rules.go [hunk 7]
```diff
@@ -39,6 +39,11 @@ func (t *egressRules) setup() error {
 
 // TODO: test negatives
 func (t *egressRules) run() error {
+	if !t.Egress {
```
Comment (ijsnellf) [inline test/integration/egress_rules.go]: @vadimeisenbergibm: please verify this is correct

File: test/integration/testdata/pilot.yaml.tmpl [hunk 8]
```diff
@@ -81,6 +81,12 @@ spec:
         args:
         - discovery
         - --v={{.Verbosity}}
+        - -a
+        - {{.Namespace}}
+        - --registries
+        - {{.Registry}}
+        - --eurekaserverURL
```
Comment (kyessenov) [inline test/integration/testdata/pilot.yaml.tmpl]: is this safe to supply if eureka mirror is not deployed?

**Comment by ijsnellf** [comment]
@kyessenov, @ayj: PTAL

File: adapter/config/ingress/BUILD [hunk 9]
```diff
@@ -42,7 +42,9 @@ go_test(
         "//proxy:go_default_library",
         "//test/mock:go_default_library",
         "//test/util:go_default_library",
+        "@com_github_golang_glog//:go_default_library",
```
[the remark this answers is not in the document]
Comment (ijsnellf) [reply adapter/config/ingress/BUILD]: nice catch

File: test/integration/testdata/pilot.yaml.tmpl (hunk 8 above)
Comment (ijsnellf) [reply test/integration/testdata/pilot.yaml.tmpl]: yes, the value isn't used

**Pull request merged**